### PR TITLE
Lookup for hns 1

### PR DESCRIPTION
**Pull request by ankitaluthra1**
### Description

- This PR is the first PR for implementing implicit-dirs changes as part of HNS support.
- This contains changes in lookupInode implementation to use get folder api instead of ListObjects in case of HNS and corresponding tests for it.
- This also covers fix for an edge case in stat cache in which same name objects and folder can co-exist in HNS bucket.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Done
2. Unit tests - Done
3. Integration tests - Done

### CODE REVIEW

File: internal/cache/metadata/stat_cache.go [hunk 1]
```diff
@@ -247,8 +247,19 @@ func (sc *statCacheBucketView) LookUpFolder(
...
 	hit, entry := sc.sharedCacheLookup(folderName, now)
+
 	if hit {
-		return hit, entry.f
+		// Adds negative entry scenario
+		if entry.f == nil && entry.m == nil {
+			return true, nil
+		}
+
+		// Adds uninitiated object scenario
```
Comment (charith87) [inline internal/cache/metadata/stat_cache.go:257]: What is uninitiated object scenario? Could we add clarity here.

File: internal/cache/metadata/stat_cache.go [hunk 2]
```diff
@@ -247,8 +247,19 @@ func (sc *statCacheBucketView) LookUpFolder(
 	now time.Time) (bool, *gcs.Folder) {
 	// Look up in the LRU cache.
 	hit, entry := sc.sharedCacheLookup(folderName, now)
+
 	if hit {
-		return hit, entry.f
+		// Adds negative entry scenario
+		if entry.f == nil && entry.m == nil {
```
Comment (charith87) [inline internal/cache/metadata/stat_cache.go:253]: Are there older tests for stat cache negative scenario and are they passing?

File: internal/fs/inode/dir.go [hunk 3]
```diff
@@ -366,10 +365,13 @@ func findExplicitFolder(ctx context.Context, bucket *gcsx.SyncerBucket, name Nam
 		return nil, fmt.Errorf("error in get folder for lookup : %w", folderErr)
 	}
 
+	folderObject := folderResult.ConvertFolderToMinObject()
+	folderObject.Name = name.objectName
```
Comment (charith87) [inline internal/fs/inode/dir.go:369]: When would the folderName and the requestedName be different and would it be possible to move it inside the convertFolderToMinObject() if its needed in more places than here.

Comment (ankitaluthra1) [reply internal/fs/inode/dir.go]: This is the test representing TestCopyDirectoryInNonExistingDirectory case when folderName and the requestedName are different.

File: internal/cache/metadata/stat_cache.go (hunk 2 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: yes *StatCacheTest* has multiple stat cache negative scenarios for both object and folders which is passing for this PR.

File: internal/cache/metadata/stat_cache.go (hunk 1 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: Updated comment to be more descriptive, also corresponding test is also updated test name (Test_ShouldReturnHitFalseWhenOnlyObjectAlreadyHasEntry) for better documentation

File: internal/fs/inode/dir.go (hunk 3 above)
Comment (ankitaluthra1) [reply internal/fs/inode/dir.go]: Moved name inside and also added test case for same.

**Review by charith87 — APPROVED** [review]
